pkg/checker/functools.go

```diff
@@ -0,0 +1,113 @@
+		// `CheckRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+func makeRemediationRunner(f interface{}, args ...interface{}) (RemediationRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `RemediationRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+// validateFunction checks whether `f` is a function, and whether
+// the passed parameters have the correct number and types.
+func validateFunction(f interface{}, args ...interface{}) error {
+	typ := reflect.TypeOf(f)
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	if kind := typ.Kind(); kind != reflect.Func {
+		return fmt.Errorf("invalid check runner, want %v, got %v", reflect.Func, kind)
+	}
+	// check output parameter type(s). The function must return an error
+	// TODO: assess if we want to also allow (value, error) as acceptable return
+	//       types. This means also changing how values are returned in
+	//       `makeCheckRunner` and `makeRemediationRunner`.
+	switch {
+	case typ.NumOut() == 1:
+		// return type must be error
+		if typ.Out(0) != errorType {
+			return fmt.Errorf("invalid return type, want %v, got %v", errorType, typ.Out(0))
+		}
+	default:
+		return fmt.Errorf("invalid number of returned parameters, want 1, got %d", typ.NumOut())
+	}
+	// check input parameter types
+	if typ.IsVariadic() {
+		if len(args) < typ.NumIn()-1 {
+			return fmt.Errorf("wrong number of parameters for variadic function, want at least %d, got %d", typ.NumIn()-1, len(args))
+		}
+		// check non-variadic arguments
+		for idx := 0; idx < typ.NumIn()-1; idx++ {
+			want, got := typ.In(idx).Kind(), reflect.TypeOf(args[idx]).Kind()
+			if want != got {
+				return fmt.Errorf("wrong non-variadic parameter type at index %d: want %v, got %v", idx, want, got)
+			}
+		}
+		// check variadic arguments
+		vargs := typ.In(typ.NumIn() - 1)
+		if vargs.Kind() != reflect.Slice {
+			return fmt.Errorf("invalid variadic arguments: want %v, got %v", reflect.Slice, vargs.Kind())
+		}
+		for idx := typ.NumIn() - 1; idx < len(args); idx++ {
+			argType := reflect.TypeOf(args[idx])
+			if argType.Kind() == reflect.Interface {
+				if !argType.Implements(vargs.Elem()) {
+					return fmt.Errorf("invalid interface type for variadic argument at index %d: %v does not implement %v", idx, reflect.TypeOf(args[idx]), vargs.Elem().Kind())
+				}
+			} else {
+				if argType.Kind() != vargs.Elem().Kind() {
+					return fmt.Errorf("invalid type for variadic argument at index %d: %v is not %v", idx, argType.Kind(), vargs.Elem().Kind())
+				}
+			}
```

Comment:
I looks like some bits of logic is missing here. There're three valid cases possible here:
1. We require specific type, and we receive specific type.
2. We require an interface, but we receive a specific type (which implements the interface).
3. We require an interface, and we receive and interface (which implements the first interface).

I see only cases 1 and 3 here, but not 2.

I think this piece of code should have a comment about it, and it should be clear where which case is solved.

Reply:
I will address this after adding unit tests